src/tools/miri/src/shims/unix/linux/sync.rs

```diff
@@ -189,6 +231,16 @@ pub fn futex<'tcx>(
             // before doing the syscall.
             this.atomic_fence(AtomicFenceOrd::SeqCst)?;
             let mut n = 0;
+
+            // Check that the waiter is waiting using FUTEX_WAIT_*.
+            if this.futex_waiter_count(addr_usize) != 0 {
+                if this.futex_top_waiter_extra(addr_usize).is_some() {
```

Comment:
Please use `&&`. Also, like above: please explain why, and make sure we have a test.

Given that this takes up N waiters, why are we only testing the top waiter?

Reply:
Ah, okay. Please document this invariant in the `FutexWaiter` type.